driver/src/test/java/org/neo4j/driver/internal/pool/ThreadCachingPoolTest.java

```diff
@@ -256,10 +268,126 @@ public void shouldRecovedDisposedItemReallocationFailing() throws Throwable
+        //from the live objects created in the background thread
+        PooledObject obj2 = pool.acquire( 10, TimeUnit.SECONDS );
+
+        //THEN
+        assertThat( obj1.id, equalTo( 0 ) );
+        assertThat( obj2.id, equalTo( 1 ) );
+        BlockingQueue<Slot<PooledObject>> liveQueue = (BlockingQueue<Slot<PooledObject>>) liveQueueGet.invoke( pool );
+        BlockingQueue<Slot<PooledObject>> disposedQueue =
+                (BlockingQueue<Slot<PooledObject>>) disposedQueueGet.invoke( pool );
+
+        assertThat( disposedQueue, hasSize( 1 ) );
+        assertThat( disposedQueue.poll( 10, TimeUnit.SECONDS ).value.id, equalTo( 0 ) );
+        assertThat( liveQueue, empty() );
+    }
+
+    @SuppressWarnings( "unchecked" )
+    @Test
+    public void shouldNotAddToLiveQueueTwice() throws Throwable
+    {
+        // Given
+        ThreadCachingPool<PooledObject>
+                pool = new ThreadCachingPool<>( 4, trackAllocator, checkInvalidateFlag, Clock.SYSTEM );
+
+        // When
+        PooledObject o1 = pool.acquire( 10, TimeUnit.SECONDS );
+        o1.release();
+        PooledObject o2 = pool.acquire( 10, TimeUnit.SECONDS );
+        o2.release();
+
+        // Then
+        BlockingQueue<Slot<PooledObject>> liveQueue = (BlockingQueue<Slot<PooledObject>>) liveQueueGet.invoke( pool );
+        assertThat(liveQueue, hasSize( 1 ));
+    }
+
+    @SuppressWarnings( "unchecked" )
+    @Test
+    public void shouldNotAddToLiveQueueTwice2() throws Throwable
+    {
+        // Given
+        ThreadCachingPool<PooledObject>
+                pool = new ThreadCachingPool<>( 4, trackAllocator, checkInvalidateFlag, Clock.SYSTEM );
+        PooledObject o1 = pool.acquire( 10, TimeUnit.SECONDS );
+        PooledObject o2 = pool.acquire( 10, TimeUnit.SECONDS );
+
+        // When
+        o1.release();
+        o2.release();
+
+        // Then
+        BlockingQueue<Slot<PooledObject>> liveQueue = (BlockingQueue<Slot<PooledObject>>) liveQueueGet.invoke( pool );
+        assertThat(liveQueue, hasSize( 1 ));
+    }
+
+    private boolean acquireInSeparateThread( final ThreadCachingPool<PooledObject> pool ) throws InterruptedException
+    {
+        final AtomicBoolean succeeded = new AtomicBoolean( true );
+        executor.execute( new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try
+                {
+                    PooledObject obj = pool.acquire( 10, TimeUnit.SECONDS );
+                    obj.release();
+                }
+                catch ( InterruptedException e )
+                {
+                    succeeded.set( false );
+                }
+            }
+        } );
+        executor.awaitTermination( 2, TimeUnit.SECONDS );
+        return succeeded.get();
+    }
+
+
+    //This is terrible hack, but I really want to keep the queues private in
+    //ThreadCachingPool
+    private static MethodHandle queueGetter( String name )
```

Comment:
Yeah I found this code, https://github.com/neo4j/neo4j/commits/3.1/community/bolt/src/main/java/org/neo4j/bolt/v1/packstream/utf8/SunMiscUTF8Encoder.java?author=jakewins, on teh internets and ran with it. Seemed legit.